### PR TITLE
Configuration from file to load DB values

### PR DESCRIPTION
**Pull request by tian2992**

### CODE REVIEW

**Comment by tian2992** [comment]
Got it working but still needs a test to check for loaded values.

**Comment by odscjames** [comment]
Hello, thanks for this. I've had a look:

1)

>  config.read('~/.config/ocdsdata/config.ini')

Wasn't working for me - it wouldn't find any of the file contents. According to https://stackoverflow.com/questions/2057045/pythons-os-makedirs-doesnt-understand-in-my-path you need to expand the path:

> config.read(os.path.expanduser('~/.config/ocdsdata/config.ini'))


2) 

When I fix the above with a ini file I copied from the Sample, I get: 

> ValueError: invalid literal for int() with base 10: "'5432'"

Also the strings come out with ''s around them. In a ini file, ints don't need quotes and strings don't strictly need them either. Given these should be simple strings, I'd be tempted to leave them off:

```
[DBHOST]
HOSTNAME = localhost
PORT = 5432
USERNAME = ocdsdata
DBNAME = ocdsdata
```

3)

The port won't actually be used. If you set a different port in the INI file and get config to print DB_URI, you'll get 

> postgresql://ocdsdata:password@localhost/ocdsdata

The 2 lines in the config Python module that build the URI:
>     database_uri = 'postgresql://{}:{}@{}/{}'.format(user, password, host, dbname)
>      database_uri = 'postgresql://{}:{}@{}/{}'.format(user, 'ocdsdata', host, dbname)

Need to be changed to include the port variable to.


4)

Can I suggest ocdsdata/maindatabase/sample_db_ocdsdata.ini is put in a different place? While it only has database settings in it at the moment, the filename of ~/.config/ocdsdata/config.ini suggests it could hold other config values in the future. I would therefore suggest it's just called sample_config.ini and put in the root folder of the repository.

5) 

As this is getting complex, It would be really good if this could be written up in the README file to.

For port, host, user, dbname, we now several ways you can set these:

  *   config file
  *   env variable
  *   default name, user, password, etc

For the password, there is also several ways this can be picked up:

  *  .pgpass file
  *   env variable
  *   default password

These can even be mixed, so you can set the user and dbname in the config and set the password in .pgpass! :smile:  

Thank you  - do let me know if you think any of that is wrong!
Talk soon,
James

**Comment by tian2992** [comment]
@odscjames thanks for your checks, I will take them into notice. Point per point comments:

4. Totally agree, I think we could use up the chance to have a general config file, and therefore it should probably be more general. I will change the required options both in the ini file and relocating it to a better position.

5. Yes, that can get quite complex, but also quite flexible. I shall add more info to the README.

**Comment by jpmckinney** [comment]
In thinking through the many ways in which to specify the database URL, for inspiration, there's a similar situation in Django where the options are:

* Set `DATABASES` in the `settings.py` config file
* If deploying to Heroku or otherwise following 12factor, set the `DATABASE_URL` environment variable, which is picked up by `settings.py` through [dj-database-url](https://github.com/kennethreitz/dj-database-url)
* Optionally, use `.pgpass` to store the password

I would tend towards not having default database parameters in the code, as this seems to be the responsibility of the deployment, not of the library.

**Comment by odscjames** [comment]
See #124 - I just realised the current code has some bugs in it! I put it in a separate issue as they were not introduced by this P.R.. However it would be good if they could be looked at as part of this? At least be aware of them as they might impact your work.

Let me know when you are happy with the code and have updated the Readme - I can take another look then. Thanks! :smile:

**Comment by odscjames** [comment]
> On that note, we might prefer DATABASE_URL to DB_URI in terms of name

Don't think I have any opinions on this, but maybe this should be a separate issue to make sure we stay on track?

> I would tend towards not having default database parameters in the code, as this seems to be the responsibility of the deployment, not of the library.

When you say "the deployment", I'm not totally sure which bit you mean? 

But actually maybe we shouldn't - imagine a scenario where the user has another database with the default credentials (by chance) for another program, and doesn't realise our program will try to use a DB. Our program would then try and make changes to existing data for another app! It's generally a rule that defaults should be safe, so maybe it should default to Nothing and the library should prompt the user in that case. 

I would again say this should be a separate issue but actually this would probably mean rewriting this work - so not sure about the best way forwards is. It would be good for people to confirm what they think the user stories for desired behaviour would be maybe?

**Comment by tian2992** [comment]
I agree on both topic, that the envvar should be renamed and that defaults should be separate, however I think a different file would suffice. 

We could bundle a default ini inside the same dir and load it via the config file hierarchy. Then if the main config file were to be missing, we could still load defaults.

**Comment by odscjames** [comment]
> Then if the main config file were to be missing, we could still load defaults.

My point is that maybe there should be no defaults. If the user has not specified any DB settings, just print a nice error and stop. What do you think about that?

**Comment by robredpath** [comment]
Right now, the two user stories that we're trying to satisfy are: 

"As an analyst working on the OCP Helpdesk, I want to be able to log into a server and get data into a database so that I can run queries to help publishers"

and

"As an analyst working on the OCP Helpdesk, I want to be able to get data into a database on my local machine so that I can run queries to help publishers". 

So, I think it's fine to rely quite heavily on our colleagues reading the docs and talking to us, and it's supremely unlikely that they will have existing databases that we'll clobber. 

I'm broadly indifferent on most of the detailed points discussed in this thread so far, but I do feel that having default credentials baked into the code is potentially dangerous - it's very easy to forget to go back and change the credentials later, and even if the data is open data, we don't want just anyone being able to access instances of this. 

> If the user has not specified any DB settings, just print a nice error and stop. 

Agreed.

**Comment by jpmckinney** [comment]
Agreed on no defaults.

For the user story for logging into a server: I'd expect that another person will have deployed/set up ocdsdata on the server, and will have written a config file as part of that. When the analyst logs in, the config file is already set up, and they can issue CLI commands without setting env vars, etc.

We can imagine that there may someday be other servers with their own config files. This is what I meant by the database parameters being the responsibility of each deployment; it's a good practice to separate config from code ([some discussion here](https://www.12factor.net/config), where the solution is to rely only on env vars).

**Comment by tian2992** [comment]
I disagree on no defaults, but seems there is a consensus against. Shall I trigger an error then if neither DBURI nor config file are set?

**Comment by odscjames** [comment]
Print a nice message for humans and then "quit(-1)". That way any calling process can check the return code and tell an error occurred.

File: sample_ocdsdata_config.ini [hunk 1]
```diff
@@ -0,0 +1,7 @@
+; sample_ocdsdata_config.ini
+[DBHOST]
+HOSTNAME = localhost
+PORT = 5432
+USERNAME = ocdsdata
+PASSWORD = ocdsdata
```
Comment (robredpath) [inline sample_ocdsdata_config.ini]: Can we change this to changeme? I can imagine being lazy and new to Postgres and just setting this as the password!

**Comment by odscjames** [comment]
Just quickly testing now ....

**Pull request merged**